launch_testing/launch_testing/pytest/plugin.py

```diff
@@ -0,0 +1,457 @@
+        # step, we look ourselves.
+        transfer_markers(obj, item.cls, item.module)
+        item = pytest.Function.from_parent(collector, name=name)  # To reload keywords.
+
+        marker = item.get_closest_marker('launch_testing')
+        if marker is not None:
+            # inject the correct launch_testing fixture here
+            if 'fixture' not in marker.kwargs:
+                warnings.warn(
+                    '"fixture" keyword argument is required in a pytest.mark.launch_testing() '
+                    f'decorator: \n{get_error_context_from_obj(item.obj)}'
+                )
+                return None
+            fixture = marker.kwargs['fixture']
+            fixturename = fixture.__name__
+            scope = fixture._pytestfixturefunction.scope
+            is_shutdown = marker.kwargs.get('shutdown', False)
+            items = generate_test_items(collector, name, obj, fixturename, is_shutdown)
+            if need_shutdown_test_item(obj) and scope != 'function':
+                # for function scope we only need one shutdown item
+                # if not we're going to use two event loops!!!
+                shutdown_items = generate_test_items(collector, name, obj, fixturename, True)
+                for item, shutdown_item in zip(items, shutdown_items):
+                    item._launch_testing_shutdown_item = shutdown_item
+                items.extend(shutdown_items)
+            return items
+
+
+def get_error_context_from_obj(obj):
+    """Return formatted information of the object location."""
+    try:
+        fspath = inspect.getsourcefile(obj)
+    except TypeError:
+        return 'location information of the object not found'
+    try:
+        lines, lineno = inspect.getsourcelines(obj)
+    except IOError:
+        return f'file {fspath}: source code not available'
+    error_msg = f'file {fspath}, line {lineno}'
+    for line in lines:
+        line = line.rstrip()
+        error_msg += f'\n  {line}'
+        if line.lstrip().startswith('def'):
+            break
+    return error_msg
+
+
+@pytest.mark.trylast
+def pytest_collection_modifyitems(session, config, items):
+    """Reorder tests, so shutdown tests happen after the corresponding fixture teardown."""
+
+    def cmp(left, right):
+        leftm = left.get_closest_marker('launch_testing')
+        rightm = right.get_closest_marker('launch_testing')
+        if None in (leftm, rightm):
+            return 0
+        fixture = leftm.kwargs['fixture']
+        if fixture is not rightm.kwargs['fixture']:
+            return 0
+        if fixture._pytestfixturefunction.scope == 'function':
+            return 0
+        left_is_shutdown = int(left._launch_testing_is_shutdown)
+        right_is_shutdown = int(right._launch_testing_is_shutdown)
+        return left_is_shutdown - right_is_shutdown
+
+    # python sort is guaranteed to be stable
+    items.sort(key=functools.cmp_to_key(cmp))
```

Comment:
I think I'm going to rewrite the sorting manually.
the currently  cmp() method doesn't respect the transivity property:

you might have a < b, b=c, c=a.
I don't think python sorting handles that correctly, I'm not sure honestly.

Reply:
I also want to double check that all tests using the same launch test fixture are already one next to the other in the items list, I think that's the case but I'm not completely sure.
If that's not the case, we need to do something more complex here.

Reply:
What pytest default hook does https://github.com/pytest-dev/pytest/blob/a83b359cf0902e0e55b79411dd54480be8a0aa4e/src/_pytest/fixtures.py#L1604 (which is run before this hook).

Reply:
done, I didn't use exactly that snippet though, see https://github.com/ros2/launch/pull/528/commits/46ba3b249c5f23869d750816bba35c67f09d4ce1